flytekit/core/promise.py

```diff
@@ -693,15 +693,23 @@ def create_and_link_node(
         retry_strategy or _literal_models.RetryStrategy(0),
     )
 
+    # Node ID should be the entity name, but we need to add a suffix to ensure unique node IDs
+    # when we encounter more than one such entity during compilation
+    node_id = (
+        f"{entity.name}-n{ctx.compilation_state.get_entity_count(entity.name)}"
```

Comment:
I think we still need the prefix... I added a test https://github.com/flyteorg/flytekit/commit/ff2a5509cc3a34c4dd089ff55fa6064f421e0540#diff-c34bbbaaada631ae99ff9734e7955d79dbc84791c54e5ee177ca796f11878109R153 that might fail weirdly now?